core/src/main/java/software/amazon/awssdk/auth/ContainerCredentialsProvider.java

```diff
@@ -0,0 +1,202 @@
+            if (!ALLOWED_HOSTS.contains(uri.getHost())) {
+
+                throw new SdkClientException(
+                    String.format("The full URI (%s) contained withing environment variable %s has an invalid host. "
+                                  + "Host can only be one of [%s].",
+                                  uri,
+                                  AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.environmentVariable(),
+                                  ALLOWED_HOSTS.stream().collect(joining(","))));
+            }
+            return uri;
+        }
+
+        @Override
+        public Map<String, String> headers() {
+            return AwsSystemSetting.AWS_CONTAINER_AUTHORIZATION_TOKEN.getStringValue()
+                                                                     .filter(StringUtils::isNotBlank)
+                                                                     .map(t -> Collections.singletonMap("Authorization", t))
+                                                                     .orElseGet(Collections::emptyMap);
+        }
+
+        @Override
+        public boolean isEnabled() {
+            return AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.getStringValue().isPresent();
+        }
+
+        @Override
+        public String disabledExceptionMessage() {
+            return String.format(
+                "Credentials cannot be loaded from the container because the credentials environment variable (%s) and system "
+                + "property (%s) are not set or cannot be accessed due to the security manager.",
+                AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.environmentVariable(),
+                AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.property());
+        }
+    }
+
+    /**
+     *
+     */
+
+    /**
+     * A builder for creating a custom a {@link ContainerCredentialsProvider}.
+     */
+    public static final class Builder {
+        private Boolean asyncCredentialUpdateEnabled = false;
+        private CredentialsEndpointProvider credentialsEndpointProvider;
+
+        /**
+         * Created using {@link #builder()}.
+         */
+        private Builder() {}
+
+        /**
+         * Configure whether this provider should fetch credentials asynchronously in the background. If this is true, threads are
+         * less likely to block when {@link #getCredentials()} is called, but additional resources are used to maintain the
+         * provider.
+         *
+         * <p>By default, this is disabled.</p>
+         */
+        public Builder asyncCredentialUpdateEnabled(Boolean asyncCredentialUpdateEnabled) {
+            this.asyncCredentialUpdateEnabled = asyncCredentialUpdateEnabled;
+            return this;
+        }
+
+        public Builder credentialsEndpointProvider(CredentialsEndpointProvider credentialsEndpointProvider) {
+            this.credentialsEndpointProvider = credentialsEndpointProvider;
+            return this;
+        }
```

Comment:
Can we document this? It's a tricky field for the customer to understand, I think.

Reply:
This is pretty internal; I've marked the whole builder as package-private